File: fsrs-java/src/main/java/dev/bee/fsrs/DefaultFsrsEngine.java
package dev.bee.fsrs;

final class DefaultFsrsEngine implements FsrsEngine {
    private static final String PARAM_RATING = "rating";

    private final FsrsParameters parameters;

    DefaultFsrsEngine(FsrsParameters parameters) {
        this.parameters = Fsrs.requireNonNull(parameters, "parameters");
    }

    @Override
    public FsrsMemoryState initialState(FsrsRating firstRating) {
        Fsrs.requireNonNull(firstRating, "firstRating");
        return new FsrsMemoryState(initialStability(firstRating), initialDifficulty(firstRating, true));
    }

    @Override
    public double retrievability(FsrsMemoryState state, int elapsedDays) {
        Fsrs.requireNonNull(state, "state");
        Fsrs.validateElapsedDays(elapsedDays);
        double stability = Math.max(state.stability(), Fsrs.STABILITY_MIN);
        return Math.pow(1.0 + parameters.factor() * elapsedDays / stability, parameters.decay());
    }

    @Override
    public FsrsMemoryState nextState(FsrsMemoryState previousState, FsrsRating rating, int elapsedDays) {
        Fsrs.requireNonNull(previousState, "previousState");
        Fsrs.requireNonNull(rating, PARAM_RATING);
        Fsrs.validateElapsedDays(elapsedDays);
        validateDifficulty(previousState.difficulty());

        double retrievability = retrievability(previousState, elapsedDays);
        double nextDifficulty = nextDifficulty(previousState.difficulty(), rating);
        double nextStability;
        if (rating == FsrsRating.AGAIN) {
            nextStability = nextForgetStability(nextDifficulty, previousState.stability(), retrievability);
        } else if (elapsedDays == 0) {
            nextStability = shortTermStability(previousState.stability(), rating);
        } else {
            nextStability = nextRecallStability(nextDifficulty, previousState.stability(), retrievability, rating);
        }
        return new FsrsMemoryState(nextStability, nextDifficulty);
    }

    @Override
    public double nextDifficulty(double currentDifficulty, FsrsRating rating) {
        validateDifficulty(currentDifficulty);
        Fsrs.requireNonNull(rating, PARAM_RATING);
        double deltaDifficulty = -(parameters.difficultyDeltaScale() * (rating.value() - 3.0));
        double linearDamping = (10.0 - currentDifficulty) * deltaDifficulty / 9.0;
        double easyInitialDifficulty = initialDifficulty(FsrsRating.EASY, false);
        double meanReversion = parameters.difficultyMeanReversionWeight();
        double next = meanReversion * easyInitialDifficulty
                + (1.0 - meanReversion) * (currentDifficulty + linearDamping);
        return Fsrs.clamp(next, Fsrs.MIN_DIFFICULTY, Fsrs.MAX_DIFFICULTY);
    }

    @Override
    public double shortTermStability(double stability, FsrsRating rating) {
        if (!Double.isFinite(stability) || stability <= 0.0) {
            throw new IllegalArgumentException("stability must be finite and positive");
        }
        Fsrs.requireNonNull(rating, PARAM_RATING);
        double increase = Math.exp(parameters.shortTermBase() * (rating.value() - 3.0 + parameters.shortTermRatingOffset()))
                * Math.pow(stability, -parameters.shortTermStabilityDecay());
        if (rating == FsrsRating.GOOD || rating == FsrsRating.EASY) {
            increase = Math.max(increase, 1.0);
        }
        return Math.max(stability * increase, Fsrs.STABILITY_MIN);
    }

    @Override
    public int nextIntervalDays(double stability, double desiredRetention, int maximumInterval) {
        if (!Double.isFinite(stability) || stability <= 0.0) {
            throw new IllegalArgumentException("stability must be finite and positive");
        }
        Fsrs.validateDesiredRetention(desiredRetention);
        Fsrs.validateMaximumInterval(maximumInterval);
        double interval = (stability / parameters.factor())
                * (Math.pow(desiredRetention, 1.0 / parameters.decay()) - 1.0);
        int rounded = (int) Math.round(interval);
        return Math.min(Math.max(rounded, 1), maximumInterval);
    }

    @Override
    public FsrsReviewOutput review(FsrsReviewInput input) {
        Fsrs.requireNonNull(input, "input");
        double retrievability = retrievability(input.previousState(), input.elapsedDays());
        FsrsMemoryState nextState = nextState(input.previousState(), input.rating(), input.elapsedDays());
        int interval = nextIntervalDays(nextState.stability(), input.desiredRetention(), input.maximumInterval());
        return new FsrsReviewOutput(nextState, retrievability, interval);
    }

    private double initialStability(FsrsRating rating) {
        return Math.max(parameters.initialStability(rating), Fsrs.STABILITY_MIN);
    }

    private double initialDifficulty(FsrsRating rating, boolean clamp) {
        double difficulty = parameters.initialDifficultyBase()
                - Math.exp(parameters.initialDifficultyExponent() * (rating.value() - 1.0))
                + 1.0;
        return clamp ? Fsrs.clamp(difficulty, Fsrs.MIN_DIFFICULTY, Fsrs.MAX_DIFFICULTY) : difficulty;
    }

    private double nextRecallStability(
            double difficulty,
            double stability,
            double retrievability,
            FsrsRating rating
    ) {
        double hardPenalty = rating == FsrsRating.HARD ? parameters.hardPenalty() : 1.0;
        double easyBonus = rating == FsrsRating.EASY ? parameters.easyBonus() : 1.0;
        double next = stability * (
                1.0
                        + Math.exp(parameters.recallStabilityBase())
                        * (11.0 - difficulty)
                        * Math.pow(stability, -parameters.recallStabilityStabilityDecay())
                        * (Math.exp((1.0 - retrievability) * parameters.recallStabilityRetrievabilitySensitivity()) - 1.0)
                        * hardPenalty
                        * easyBonus
        );
        return Math.max(next, Fsrs.STABILITY_MIN);
    }

    private double nextForgetStability(double difficulty, double stability, double retrievability) {
        double longTermForget = parameters.forgetStabilityBase()
                * Math.pow(difficulty, -parameters.forgetStabilityDifficultyDecay())
                * (Math.pow(stability + 1.0, parameters.forgetStabilityStabilityGrowth()) - 1.0)
                * Math.exp((1.0 - retrievability) * parameters.forgetStabilityRetrievabilitySensitivity());
        double shortTermForgetCap = stability / Math.exp(parameters.shortTermBase() * parameters.shortTermRatingOffset());
        return Math.max(Math.min(longTermForget, shortTermForgetCap), Fsrs.STABILITY_MIN);
    }

    private static void validateDifficulty(double difficulty) {
        if (!Double.isFinite(difficulty) || difficulty < Fsrs.MIN_DIFFICULTY || difficulty > Fsrs.MAX_DIFFICULTY) {
            throw new IllegalArgumentException("difficulty must be finite and in [1, 10]");
        }
    }
}

File: bee-fsrs/src/main/kotlin/dev/bee/fsrs/DefaultFsrsEngine.kt
package dev.bee.fsrs

internal class DefaultFsrsEngine(
    parameters: FsrsParameters?,
) : FsrsEngine {
    private val parameters: FsrsParameters = Fsrs.requireNonNull(parameters, "parameters")

    override fun initialState(firstRating: FsrsRating?): FsrsMemoryState {
        val rating = Fsrs.requireNonNull(firstRating, "firstRating")
        return FsrsMemoryState(initialStability(rating), initialDifficulty(rating, clamp = true))
    }

    override fun retrievability(state: FsrsMemoryState?, elapsedDays: Int): Double {
        val resolvedState = Fsrs.requireNonNull(state, "state")
        Fsrs.validateElapsedDays(elapsedDays)
        val stability = resolvedState.stability.coerceAtLeast(Fsrs.STABILITY_MIN)
        return Math.pow(1.0 + parameters.factor() * elapsedDays / stability, parameters.decay())
    }

    override fun nextState(
        previousState: FsrsMemoryState?,
        rating: FsrsRating?,
        elapsedDays: Int,
    ): FsrsMemoryState {
        val resolvedState = Fsrs.requireNonNull(previousState, "previousState")
        val resolvedRating = Fsrs.requireNonNull(rating, PARAM_RATING)
        Fsrs.validateElapsedDays(elapsedDays)
        validateDifficulty(resolvedState.difficulty)

        val retrievability = retrievability(resolvedState, elapsedDays)
        val nextDifficulty = nextDifficulty(resolvedState.difficulty, resolvedRating)
        val nextStability = when {
            // py-fsrs v6.3.1 routes every same-day review, including Again,
            // through the short-term stability update; forget stability is
            // only used for reviews at least one day out.
            elapsedDays == 0 -> shortTermStability(resolvedState.stability, resolvedRating)
            resolvedRating == FsrsRating.AGAIN -> nextForgetStability(
                nextDifficulty,
                resolvedState.stability,
                retrievability,
            )

            else -> nextRecallStability(nextDifficulty, resolvedState.stability, retrievability, resolvedRating)
        }
        return FsrsMemoryState(nextStability, nextDifficulty)
    }

    override fun nextDifficulty(currentDifficulty: Double, rating: FsrsRating?): Double {
        validateDifficulty(currentDifficulty)
        val resolvedRating = Fsrs.requireNonNull(rating, PARAM_RATING)
        val deltaDifficulty = -(parameters.difficultyDeltaScale() * (resolvedRating.value() - 3.0))
        val linearDamping = (10.0 - currentDifficulty) * deltaDifficulty / 9.0
        val easyInitialDifficulty = initialDifficulty(FsrsRating.EASY, clamp = false)
        val meanReversion = parameters.difficultyMeanReversionWeight()
        val next = meanReversion * easyInitialDifficulty +
            (1.0 - meanReversion) * (currentDifficulty + linearDamping)
        return Fsrs.clamp(next, Fsrs.MIN_DIFFICULTY, Fsrs.MAX_DIFFICULTY)
    }

    override fun shortTermStability(stability: Double, rating: FsrsRating?): Double {
        require(stability.isFinite() && stability > 0.0) { "stability must be finite and positive" }
        val resolvedRating = Fsrs.requireNonNull(rating, PARAM_RATING)
        var increase = Math.exp(parameters.shortTermBase() * (resolvedRating.value() - 3.0 + parameters.shortTermRatingOffset())) *
            Math.pow(stability, -parameters.shortTermStabilityDecay())
        if (resolvedRating == FsrsRating.GOOD || resolvedRating == FsrsRating.EASY) {
            increase = increase.coerceAtLeast(1.0)
        }
        return (stability * increase).coerceAtLeast(Fsrs.STABILITY_MIN)
    }

    override fun nextIntervalDays(stability: Double, desiredRetention: Double, maximumInterval: Int): Int {
        require(stability.isFinite() && stability > 0.0) { "stability must be finite and positive" }
        Fsrs.validateDesiredRetention(desiredRetention)
        Fsrs.validateMaximumInterval(maximumInterval)
        val interval = (stability / parameters.factor()) *
            (Math.pow(desiredRetention, 1.0 / parameters.decay()) - 1.0)
        // Clamp in Long before narrowing so extreme stabilities cannot wrap
        // Int and collapse to the one-day minimum.
        return Math.round(interval).coerceIn(1L, maximumInterval.toLong()).toInt()
    }

    override fun review(input: FsrsReviewInput?): FsrsReviewOutput {
        val resolvedInput = Fsrs.requireNonNull(input, "input")
        val previousState = Fsrs.requireNonNull(resolvedInput.previousState, "previousState")
        val rating = Fsrs.requireNonNull(resolvedInput.rating, "rating")
        val retrievability = retrievability(previousState, resolvedInput.elapsedDays)
        val nextState = nextState(previousState, rating, resolvedInput.elapsedDays)
        val interval = nextIntervalDays(nextState.stability, resolvedInput.desiredRetention, resolvedInput.maximumInterval)
        return FsrsReviewOutput(nextState, retrievability, interval)
    }

    private fun initialStability(rating: FsrsRating): Double =
        parameters.initialStability(rating).coerceAtLeast(Fsrs.STABILITY_MIN)

    private fun initialDifficulty(rating: FsrsRating, clamp: Boolean): Double {
        val difficulty = parameters.initialDifficultyBase() -
            Math.exp(parameters.initialDifficultyExponent() * (rating.value() - 1.0)) +
            1.0
        return if (clamp) {
            Fsrs.clamp(difficulty, Fsrs.MIN_DIFFICULTY, Fsrs.MAX_DIFFICULTY)
        } else {
            difficulty
        }
    }

    private fun nextRecallStability(
        difficulty: Double,
        stability: Double,
        retrievability: Double,
        rating: FsrsRating,
    ): Double {
        val hardPenalty = if (rating == FsrsRating.HARD) parameters.hardPenalty() else 1.0
        val easyBonus = if (rating == FsrsRating.EASY) parameters.easyBonus() else 1.0
        val next = stability * (
            1.0 +
                Math.exp(parameters.recallStabilityBase()) *
                (11.0 - difficulty) *
                Math.pow(stability, -parameters.recallStabilityStabilityDecay()) *
                (Math.exp((1.0 - retrievability) * parameters.recallStabilityRetrievabilitySensitivity()) - 1.0) *
                hardPenalty *
                easyBonus
            )
        return next.coerceAtLeast(Fsrs.STABILITY_MIN)
    }

    private fun nextForgetStability(
        difficulty: Double,
        stability: Double,
        retrievability: Double,
    ): Double {
        val longTermForget = parameters.forgetStabilityBase() *
            Math.pow(difficulty, -parameters.forgetStabilityDifficultyDecay()) *
            (Math.pow(stability + 1.0, parameters.forgetStabilityStabilityGrowth()) - 1.0) *
            Math.exp((1.0 - retrievability) * parameters.forgetStabilityRetrievabilitySensitivity())
        val shortTermForgetCap = stability / Math.exp(parameters.shortTermBase() * parameters.shortTermRatingOffset())
        return longTermForget.coerceAtMost(shortTermForgetCap).coerceAtLeast(Fsrs.STABILITY_MIN)
    }

    private companion object {
        private const val PARAM_RATING = "rating"

        private fun validateDifficulty(difficulty: Double) {
            require(
                difficulty.isFinite() &&
                    difficulty >= Fsrs.MIN_DIFFICULTY &&
                    difficulty <= Fsrs.MAX_DIFFICULTY
            ) {
                "difficulty must be finite and in [1, 10]"
            }
        }
    }
}
